drivers/usb/device/usb_dc_mcux_lpcip3511.c

```diff
@@ -0,0 +1,775 @@
+/*
+ * Copyright (c) 2018-2019, NXP
+ * Copyright (c) 2019 PHYTEC Messtechnik GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_lpc_usbd
+
+#include <kernel.h>
+#include <soc.h>
+#include <string.h>
+#include <drivers/usb/usb_dc.h>
+#include <usb/usb_device.h>
+#include <soc.h>
+#include <device.h>
+#include "usb_dc_mcux.h"
+#include "usb_device_lpcip3511.h"
+
+#define LOG_LEVEL CONFIG_USB_DRIVER_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(usb_dc_mcux_lpcip3511);
+
+#define CONTROLLER_ID kUSB_ControllerLpcIp3511Hs0
+
+static void usb_isr_handler(void);
+extern void USB_DeviceLpcIp3511IsrFunction(void *deviceHandle);
+
+/* the setup transfer state */
+#define SETUP_DATA_STAGE_DONE	(0)
+#define SETUP_DATA_STAGE_IN	(1)
+#define SETUP_DATA_STAGE_OUT	(2)
+
+/*
+ * Endpoint absolute index calculation:
+ *
+ * MCUX EHCI USB device controller supports a specific
+ * number of bidirectional endpoints. Bidirectional means
+ * that an endpoint object is represented to the outside
+ * as an OUT and an IN Eindpoint with its own buffers
+ * and control structures.
+ *
+ * EP_ABS_IDX refers to the corresponding control
+ * structure, for example:
+ *
+ *  EP addr | ep_idx | ep_abs_idx
+ * -------------------------------
+ *  0x00    | 0x00   | 0x00
+ *  0x80    | 0x00   | 0x01
+ *  0x01    | 0x01   | 0x02
+ *  0x81    | 0x01   | 0x03
+ *  ....    | ....   | ....
+ *
+ * The NUM_OF_EP_MAX (and number of s_ep_ctrl) should be double
+ * of num_bidir_endpoints.
+ */
+#define EP_ABS_IDX(ep)		(USB_EP_GET_IDX(ep) * 2 + \
+					(USB_EP_GET_DIR(ep) >> 7))
+#define NUM_OF_EP_MAX           (DT_INST_PROP(0, num_bidir_endpoints) * 2)
+
+/* The minimum value is 1 */
+#define EP_BUF_NUMOF_BLOCKS	((NUM_OF_EP_MAX + 3) / 4)
+
+/* The max MPS is 1023 for FS, 1024 for HS.
+ * IP3511 controller need the 64-bytes aligned buffer.
+ */
+K_MEM_SLAB_DEFINE(ep_buf_pool, 1024, EP_BUF_NUMOF_BLOCKS, 64);
+
+static usb_ep_ctrl_data_t s_ep_ctrl[NUM_OF_EP_MAX];
+static usb_device_struct_t dev_data;
+
+/* LPCIP3511 device driver interface */
+static const usb_device_controller_interface_struct_t ip3511 = {
+		USB_DeviceLpc3511IpInit, USB_DeviceLpc3511IpDeinit, USB_DeviceLpc3511IpSend,
+		USB_DeviceLpc3511IpRecv, USB_DeviceLpc3511IpCancel, USB_DeviceLpc3511IpControl
+};
```

Comment:
Would this interface abstraction make it possible to merge this driver with the existing MCUX EHCI driver?

Reply:
I have merged the two drivers. I have also renamed `usb_dc_mcux_ehci.c`  to `usb_dc_mcux.c` to reflect this change.